rust/lance/src/io/commit.rs

```diff
@@ -246,28 +247,34 @@ pub(crate) async fn migrate_fragments(
 ///
 /// Indices might be missing `fragment_bitmap`, so this function will add it.
 async fn migrate_indices(dataset: &Dataset, indices: &mut [Index]) -> Result<()> {
-    // Early return so we have a fast path if they are already migrated.
-    if indices.iter().all(|i| i.fragment_bitmap.is_some()) {
-        return Ok(());
-    }
-
     for index in indices {
-        if index.fragment_bitmap.is_none() {
-            // Load the read version of the index
-            let old_version = match dataset.checkout_version(index.dataset_version).await {
-                Ok(dataset) => dataset,
-                // If the version doesn't exist anymore, skip it.
-                Err(crate::Error::DatasetNotFound { .. }) => continue,
-                // Any other error we return.
-                Err(e) => return Err(e),
+        let mut must_recalculate_fragment_bitmap = index.fragment_bitmap.is_none();
+        if !must_recalculate_fragment_bitmap {
+            match dataset.checkout_version(index.dataset_version).await {
+                Ok(old_dataset) => {
+                    must_recalculate_fragment_bitmap |=
+                        if let Some(writer_version) = &old_dataset.manifest.writer_version {
+                            writer_version.older_than(0, 8, 15)
+                        } else {
+                            true
+                        };
+                }
+                Err(crate::Error::NotFound { .. }) | Err(crate::Error::DatasetNotFound { .. }) => {
+                    must_recalculate_fragment_bitmap = true;
+                }
+                Err(e) => {
+                    return Err(e);
```

Comment:
This is unfortunate because we now have to pay this cost on every commit.  Fortunately, it is probably fairly rare that commits involve indices (and those operations are probably already kind of slow).

Perhaps someday in the future, when we feel confident no one is using these old versions, we can pull this check back out.